dhcpv4/nclient4/client.go

```diff
@@ -478,6 +478,21 @@ func (c *Client) Request(ctx context.Context, modifiers ...dhcpv4.Modifier) (lea
 	return c.RequestFromOffer(ctx, offer, modifiers...)
 }
 
+// Renew sends a lease renewal based on the given offer. RENEW requests only have a minor set of
+// changes to their options compared to SELECT requests as specified by RFC 2131, section 4.3.2.
+func (c *Client) Renew(ctx context.Context, offer *dhcpv4.DHCPv4, modifiers ...dhcpv4.Modifier) (*Lease, error) {
```

Comment:
The reasoning is that it only requires information from the offer and not the ACK. Taking in a narrower interface allows for better flexibility. For example in Talos we only store the offer from of a lease as the ACK information is parsed right away (see [here](https://github.com/siderolabs/talos/pull/5897/files#diff-009d3627b4be29dee368cb384b84645b5cd590b8eb1febfd1e548599c0bb6d2fR529-R536)). If one has a full lease, they can still very natually call `Renew(ctx, lease.offer)`, as it makes sense to renew the lease based on the information that was initially offered to a client (such as the original IP).

Since `Renew` takes modifiers and is more closely related to `Request` et al. (since it returns a `Lease`), it made sense to assimilate with those functions instead of `Release`, which works quite differently. I consider `Request` and `Renew` to be a pair, for example in Talos the same procedure is used to call both.

Reply:
This approach turned out to be insufficient with DHCP servers that return incomplete ACKs during lease renewals (so the lease cannot be rotated), so I have now changed `Renew` to live next to and work just like `Release` and take in the entire lease. I managed to do this in a relatively clean way that also allowed for some additional unit testing of `NewRenewFromOffer`.

Reply:
The new `Renew` approach is in https://github.com/insomniacslk/dhcp/pull/469/commits/aeb78d02bd032611653400cd55e3331a1f839aa9 and the corresponding unit tests in https://github.com/insomniacslk/dhcp/pull/469/commits/feb14de874aeca08e22d0d0a2f7859315646289d.